lib/msf/ui/debug.rb

```diff
@@ -0,0 +1,253 @@
+          Please provide the below information in any Github issues you open. New issues can be opened here #{get_issue_link}
+          %red%undENSURE YOU HAVE REMOVED ANY SENSITIVE INFORMATION BEFORE SUBMITTING!%clr
+
+          ===8<=== CUT AND PASTE EVERYTHING BELOW THIS LINE ===8<===
+
+
+        PREMABLE
+      end
+
+      def self.all(framework, driver)
+        all_information = preamble
+        all_information += datastore(framework, driver)
+        all_information += history
+        all_information += errors
+        all_information += logs
+        all_information += versions(framework)
+
+        all_information
+      end
+
+      def self.datastore(framework, driver)
+        # Generate an ini with the existing config file
+        ini = Rex::Parser::Ini.new(Msf::Config.config_file)
+
+        # Delete all groups from the config ini that potentially have more up to date information
+        ini.keys.each do |k|
+          unless k =~ %r{^framework/database}
+            ini.delete(k)
+          end
+        end
+
+        # Retrieve and add more up to date information
+        add_hash_to_ini_group(ini, framework.datastore, driver.get_config_core)
+        add_hash_to_ini_group(ini, driver.get_config, driver.get_config_group)
+
+        if driver.active_module
+          ds = driver.active_module.datastore.dup
+
+          # Ensures that the local default value of a variable isn't set if a global value of a variable is set
+          ds.keys.each do |k|
+            # Removes an active module datastore item if it has a default value & it matches the default value
+            if driver.active_module.options.key?(k.upcase) && driver.active_module.options[k.upcase].default == ds[k]
+              ds.delete(k)
+            end
+          end
+
+          add_hash_to_ini_group(ini, ds, driver.active_module.refname)
+        end
+
+        if ini.to_s.empty?
+          content = 'The local config file is empty, no global variables are set, and there is no active module.'
+        else
+          content = ini.to_s
+        end
+
+        build_section('Module/Datastore',
+                      'The following global/module datastore, & databse setup was configured before the issue occurred:',
+                      content)
+      end
+
+      def self.history
+        end_pos = Readline::HISTORY.length - 1
+        start_pos = end_pos > COMMAND_HISTORY_TOTAL ? end_pos - (COMMAND_HISTORY_TOTAL - 1) : 0
+
+        commands = ''
+        while start_pos <= end_pos
+          # Formats command position in history to 6 characters in length
+          commands += "#{'%-6.6s' % start_pos.to_s} #{Readline::HISTORY[start_pos]}\n"
+          start_pos += 1
+        end
+
+        build_section('History',
+                      'The following commands were ran before this issue occurred:',
+                      commands)
+      end
+
+      def self.errors
+        errors = File.read(Msf::Config.log_directory + File::SEPARATOR + 'error.log')
```

Comment:
`File.join`

Reply:
We might be able to test this with by creating a new fixtures folder with example logs, then mocking the log directory path:

```ruby
require '...'

describe Msf::Ui::Debug
  let(:fake_log_directory) { File.join(FILE_FIXTURES_PATH, "fake_logs") }

  before(:each) do
    allow(Msf::Config).to receive(:log_directory).and_return(fake_log_directory)
  end

  it "correctly parses the error files" do
    expect(subject.errors).to eql("...")
  end
end
```